shared/src/question-generators/heap/heapGenerator.ts

```diff
@@ -0,0 +1,357 @@
+import {
+  FreeTextFeedbackFunction,
+  FreeTextFormatFunction,
+  FreeTextQuestion,
+  minimalMultipleChoiceFeedback,
+  MultipleChoiceQuestion,
+  QuestionGenerator,
+} from "@shared/api/QuestionGenerator.ts"
+import { serializeGeneratorCall } from "@shared/api/QuestionRouter.ts"
+import { MaxHeap, MinHeap } from "@shared/question-generators/heap/heapMinMax.ts"
+import {
+  generateHeapsForQuestion,
+  generateNeighbourOptions,
+  generateOperationSequence,
+} from "@shared/question-generators/heap/utils.ts"
+import Random from "@shared/utils/random.ts"
+import { t, tFunctional, Translations } from "@shared/utils/translations.ts"
+
+// TODO improve both descriptions translations of the generators
+const translationHeapOperations: Translations = {
+  en: {
+    name: "Heap-Operations",
+    description: "Determine the state of a Heap after various operations",
+    taskInsert:
+      "Consider a **{{0}}-Heap**. Initally the Heap is empty. You insert the following elements in the given order: {{1}} Enter the state of the Heap after the insertions.",
+    taskExtract:
+      "Consider the following **{{0}}-Heap**: {{1}} You extract the {{2}} element **{{3}}** times. Enter the state of the Heap after the extractions.",
+    taskBuild:
+      "Consider the following Array: {{0}} Enter the state of the Heap after **Build-{{1}}-Heap** on the above array.",
+    taskCombine:
+      "Let **{{0}}** be a sequence of operations, where a **number** represents **inserting** this number into the **{{2}}-Heap** and $*$ represents an **Extract-{{2}}** operation. \\[{{0}}={{1}}\\] Enter the state of the Heap after all operations. Initially the Heap is empty.",
+    checkFormat: "Please only enter Integer seperated by commas.",
+  },
+  de: {
+    name: "Heap-Operationen",
+    description: "Bestimme den Zustand eines Heaps nach unterschiedlichen Operationen",
+    taskInsert:
+      "Betrachte einen **{{0}}-Heap**. Anfänglich ist der Heap leer. Du fügst die folgenden Elemente in gegebener Reihenfolge ein: {{1}} Gib den Heap nach den Einfügeoperationen an.",
+    taskExtract:
+      "Betrachte den folgenden **{{0}}-Heap**: {{1}} Du extrahierst das {{2}} Element **{{3}}** mal. Gib den Heap nach den Extraktionen an.",
+    taskBuild:
+      "Betrachte das folgende Array: {{0}} Gib den Heap nach **Build-{{1}}-Heap** auf dem obigen Array an.",
+    taskCombine:
+      "Sei **{{0}}** eine Sequenz von Operationen, wobei eine **Zahl** das **Einfügen** dieser Zahl in den **{{2}}-Heap** repräsentiert und $*$ eine **Extract-{{2}}** Operation. \\[{{0}}={{1}}\\] Gib den Heap nach allen Operationen an. Der Heap ist anfangs leer.",
+    checkFormat: "Bitte nur ganze Zahlen durch Komma getrennt eingeben.",
+  },
+}
+
+const translationsHeapUnderstanding: Translations = {
+  en: {
+    name: "Heap-Understanding",
+    description: "Understand the properties of heaps",
```

Comment:
`Understand the properties of heaps` ist kein gültiges Lernziel, es muss ein messbares Verhalten sein. Vielleicht stattdessen `Verify basic heap properties.`